File: DSA - Collge Lab/topological_sort.c
#include<stdio.h>
int arr[25][25] = {};
int visited[25] = {};
int stack[25];
int sIndex = -1;

void sort(int v, int n){
    visited[v] = 1;
    for(int i=n;i>=0;i--){
        if(arr[v][i] && (visited[i] == 0)){
            sort(i, n);
        }
    }
    for(int i=0;i<=sIndex;i++){
        if(stack[i] == v+1) return;
    }
    stack[++sIndex] = v+1;
}

int main(){
    int n;
    scanf("%d",&n);
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
            scanf("%d",&arr[i][j]);
        }
    }
    for(int i=0;i<n;i++) sort(i, n);
    printf("The topological order is:\n");
    for(int i=sIndex; i>= 0; i--) printf("%d ", stack[i]);
}

File: GeeksForGeeks/Topological sort.java
class Solution {
    // Function to return list containing vertices in Topological order.
    static ArrayList<Integer> topologicalSort(ArrayList<ArrayList<Integer>> adj) {
        int n = adj.size();
        Stack<Integer> stk = new Stack<>();
        boolean[] visited = new boolean[n];
        
        for(int i=0;i<n;i++) {
            if(!visited[i]) topoSort(adj, stk, visited, i);
        }
        
        ArrayList<Integer> ans = new ArrayList<>(n);
        while(!stk.isEmpty()) ans.add(stk.pop());
        return ans;
    }
    
    static void topoSort(ArrayList<ArrayList<Integer>> adj, Stack<Integer> stk, boolean[] visited, int idx) {
        visited[idx] = true;
        for(Integer i : adj.get(idx)) {
            if(!visited[i]) topoSort(adj, stk, visited, i);
        }
        stk.push(idx);
    }
}
